File: Real O Who/MarketplaceAuthService.swift
import CryptoKit
import Foundation

nonisolated struct MarketplaceAuthRegistration: Sendable {
    var name: String
    var email: String
    var password: String
    var role: UserRole
    var suburb: String
}

nonisolated struct MarketplaceAuthSession: Sendable {
    var user: UserProfile
    var account: LocalAuthAccount
}

protocol MarketplaceAuthServing: Sendable {
    nonisolated func signIn(
        email: String,
        password: String,
        accounts: [LocalAuthAccount],
        users: [UserProfile]
    ) async throws -> MarketplaceAuthSession

    nonisolated func createAccount(
        registration: MarketplaceAuthRegistration,
        existingAccounts: [LocalAuthAccount]
    ) async throws -> MarketplaceAuthSession
}

nonisolated struct LocalMarketplaceAuthService: MarketplaceAuthServing, Sendable {
    nonisolated init() {}

    nonisolated func signIn(
        email: String,
        password: String,
        accounts: [LocalAuthAccount],
        users: [UserProfile]
    ) async throws -> MarketplaceAuthSession {
        let normalizedEmail = Self.normalizedEmail(email)
        guard Self.isValidEmail(normalizedEmail) else {
            throw MarketplaceAuthError.invalidEmail
        }

        guard let account = accounts.first(where: { $0.email == normalizedEmail }) else {
            throw MarketplaceAuthError.accountNotFound
        }

        guard let salt = Data(base64Encoded: account.passwordSaltBase64),
              let storedHash = Data(base64Encoded: account.passwordHashBase64) else {
            throw MarketplaceAuthError.accountUnavailable
        }

        let passwordHash = Self.hashedPassword(password, salt: salt)
        guard passwordHash == storedHash else {
            throw MarketplaceAuthError.incorrectPassword
        }

        guard let user = users.first(where: { $0.id == account.userID }) else {
            throw MarketplaceAuthError.accountUnavailable
        }

        var updatedAccount = account
        updatedAccount.lastSignedInAt = .now

        return MarketplaceAuthSession(user: user, account: updatedAccount)
    }

    nonisolated func createAccount(
        registration: MarketplaceAuthRegistration,
        existingAccounts: [LocalAuthAccount]
    ) async throws -> MarketplaceAuthSession {
        let trimmedName = registration.name.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedSuburb = registration.suburb.trimmingCharacters(in: .whitespacesAndNewlines)
        let normalizedEmail = Self.normalizedEmail(registration.email)

        guard trimmedName.split(separator: " ").count >= 2 else {
            throw MarketplaceAuthError.invalidName
        }

        guard Self.isValidEmail(normalizedEmail) else {
            throw MarketplaceAuthError.invalidEmail
        }

        guard trimmedSuburb.count >= 2 else {
            throw MarketplaceAuthError.invalidSuburb
        }

        guard registration.password.count >= 8 else {
            throw MarketplaceAuthError.weakPassword
        }

        guard !existingAccounts.contains(where: { $0.email == normalizedEmail }) else {
            throw MarketplaceAuthError.emailTaken
        }

        let user = UserProfile(
            id: UUID(),
            name: trimmedName,
            role: registration.role,
            suburb: trimmedSuburb,
            headline: registration.role == .seller
                ? "Selling privately and keeping more of the final sale."
                : "Looking to buy directly from owners without agent friction.",
            verificationNote: registration.role == .seller
                ? "Private seller account created on this device"
                : "Buyer account created on this device",
            buyerStage: registration.role == .buyer ? .browsing : nil
        )

        let salt = Data((0..<16).map { _ in UInt8.random(in: 0...255) })
        let passwordHash = Self.hashedPassword(registration.password, salt: salt)
        let account = LocalAuthAccount(
            id: UUID(),
            userID: user.id,
            email: normalizedEmail,
            passwordSaltBase64: salt.base64EncodedString(),
            passwordHashBase64: passwordHash.base64EncodedString(),
            createdAt: .now,
            lastSignedInAt: .now
        )

        return MarketplaceAuthSession(user: user, account: account)
    }

    private static func normalizedEmail(_ email: String) -> String {
        email
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
    }

    private static func isValidEmail(_ email: String) -> Bool {
        email.contains("@") && email.contains(".")
    }

    private static func hashedPassword(_ password: String, salt: Data) -> Data {
        let combined = salt + Data(password.utf8)
        let digest = SHA256.hash(data: combined)
        return Data(digest)
    }
}

File: android/app/src/main/java/com/realowho/app/auth/MarketplaceAuthService.kt
package com.realowho.app.auth

import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64
import java.util.UUID

interface MarketplaceAuthService {
    suspend fun signIn(
        email: String,
        password: String,
        accounts: List<LocalAuthAccount>,
        users: List<MarketplaceUserProfile>
    ): MarketplaceAuthSession

    suspend fun createAccount(
        registration: MarketplaceAuthRegistration,
        existingAccounts: List<LocalAuthAccount>
    ): MarketplaceAuthSession
}

class LocalMarketplaceAuthService : MarketplaceAuthService {
    private val secureRandom = SecureRandom()

    override suspend fun signIn(
        email: String,
        password: String,
        accounts: List<LocalAuthAccount>,
        users: List<MarketplaceUserProfile>
    ): MarketplaceAuthSession {
        val normalizedEmail = normalizeEmail(email)
        if (!isValidEmail(normalizedEmail)) {
            throw MarketplaceAuthException.InvalidEmail
        }

        val account = accounts.firstOrNull { it.email == normalizedEmail }
            ?: throw MarketplaceAuthException.AccountNotFound

        val salt = decodeBase64(account.passwordSaltBase64)
            ?: throw MarketplaceAuthException.AccountUnavailable
        val storedHash = decodeBase64(account.passwordHashBase64)
            ?: throw MarketplaceAuthException.AccountUnavailable
        val passwordHash = hashPassword(password, salt)

        if (!passwordHash.contentEquals(storedHash)) {
            throw MarketplaceAuthException.IncorrectPassword
        }

        val user = users.firstOrNull { it.id == account.userId }
            ?: throw MarketplaceAuthException.AccountUnavailable

        return MarketplaceAuthSession(
            user = user,
            account = account.copy(lastSignedInAt = System.currentTimeMillis())
        )
    }

    override suspend fun createAccount(
        registration: MarketplaceAuthRegistration,
        existingAccounts: List<LocalAuthAccount>
    ): MarketplaceAuthSession {
        val trimmedName = registration.name.trim()
        val trimmedSuburb = registration.suburb.trim()
        val normalizedEmail = normalizeEmail(registration.email)

        if (trimmedName.split(Regex("\\s+")).filter { it.isNotBlank() }.size < 2) {
            throw MarketplaceAuthException.InvalidName
        }

        if (!isValidEmail(normalizedEmail)) {
            throw MarketplaceAuthException.InvalidEmail
        }

        if (trimmedSuburb.length < 2) {
            throw MarketplaceAuthException.InvalidSuburb
        }

        if (registration.password.length < 8) {
            throw MarketplaceAuthException.WeakPassword
        }

        if (existingAccounts.any { it.email == normalizedEmail }) {
            throw MarketplaceAuthException.EmailTaken
        }

        val now = System.currentTimeMillis()
        val user = MarketplaceUserProfile(
            id = UUID.randomUUID().toString(),
            name = trimmedName,
            role = registration.role,
            suburb = trimmedSuburb,
            headline = if (registration.role == UserRole.SELLER) {
                "Selling privately and keeping more of the final sale."
            } else {
                "Looking to buy directly from owners without agent friction."
            },
            verificationNote = if (registration.role == UserRole.SELLER) {
                "Private seller account created on this device"
            } else {
                "Buyer account created on this device"
            },
            buyerStage = if (registration.role == UserRole.BUYER) "browsing" else null,
            createdAt = now
        )

        val salt = ByteArray(16).also(secureRandom::nextBytes)
        val passwordHash = hashPassword(registration.password, salt)
        val account = LocalAuthAccount(
            id = UUID.randomUUID().toString(),
            userId = user.id,
            email = normalizedEmail,
            passwordSaltBase64 = Base64.getEncoder().encodeToString(salt),
            passwordHashBase64 = Base64.getEncoder().encodeToString(passwordHash),
            createdAt = now,
            lastSignedInAt = now
        )

        return MarketplaceAuthSession(user = user, account = account)
    }

    private fun normalizeEmail(email: String): String = email.trim().lowercase()

    private fun isValidEmail(email: String): Boolean = email.contains("@") && email.contains(".")

    private fun decodeBase64(value: String): ByteArray? {
        return runCatching { Base64.getDecoder().decode(value) }.getOrNull()
    }

    private fun hashPassword(password: String, salt: ByteArray): ByteArray {
        return MessageDigest
            .getInstance("SHA-256")
            .digest(salt + password.toByteArray(Charsets.UTF_8))
    }
}
